Add kolibri daemon

dylanmccall opened the pull request:
This splits the Kolibri service management code into a separate d-bus daemon which communicates with the front-end application asynchronously. This solves some synchronization issues caused by running Kolibri inside of the front-end (and tying it directly to its lifecycle). The Kolibri daemon starts Kolibri automatically when it is communicated with, and it stops it automatically after a period of inactivity.

In addition, this branch (at the moment rather crudely) tweaks Kolibri slightly so that it can be started with an ephemeral port number (sourcing learningequality/kolibri#7714), which makes the service more reliable in a context where it may be stopped and started repeatedly and where it is unimportant that we use a consistent port number.

This work is also important for adding a way to run a single Kolibri instance per system, instead of per user, and it is a step towards supporting external storage devices in the Kolibri flatpak: the d-bus interface provides a clean surface we can use to pass around information about files.

Additional steps to be done:

- [x] Rebase to organize these rather ugly commits.
- [x] Add a mode where kolibri-daemon uses the system bus instead of the session bus. Ideally we'd use an environment variable here so it is possible to switch Kolibri to system-wide mode using a flatpak override.
- [x] Add a system bus mode for the search provider as well. Unlike kolibri-daemon, we will still need to run a search provider on the *session* bus, but we can teach it to proxy its system bus instance if one exists.
- [x] Clean up the weird way we're patching Kolibri, either with some targeted monkey patching or by getting that pull request cleaned up and merged upstream.
- [x] Decide if we still need the Kolibri status monitor code. I suspect not, but keeping it may reduce the delta with other platforms - or we might choose to make the status monitor platform-specific, just as managing Kolibri is.
- [x] Add an interface to get Kolibri's home directory path via kolibri-daemon and use that for the "open Kolibri home directory" menu action.
- [x] Replace synchronous dbus calls with asynchronous ones, or put them in a separate thread.
- [x] Untangle any remaining Kolibri-specific things in utils.py. We should get everything through the daemon.
- [x] Fix the FIXME in `__open_window_for_kolibri_scheme_uri`

dylanmccall committed:
Add a separate Kolibri daemon application

This application will be a dbus service that starts and stops Kolibri
and communicates relevant information with local desktop applications.

dylanmccall committed:
Talk to kolibri-daemon in kolibri-gnome

dylanmccall committed:
Add dbus properties and change notifications

dylanmccall committed:
Tidy up application code, remove service monitor

dylanmccall committed:
Move home template code to KolibriServiceSetupProcess

dylanmccall committed:
Remove KOLIBRI_HOME from globals.py

Instead, get the path from the Kolibri d-bus daemon.

dylanmccall committed:
Configure logging in each KolibriDaemon thread

Fixes #10

dylanmccall committed:
Allow accessing KolibriDaemon over the system bus

dylanmccall committed:
Add a search interface to KolibriDaemon

danigm commented on build-aux/flatpak/python3-kolibri.json:
```diff
@@ -0,0 +1,24 @@
+    "buildsystem": "simple",
+    "build-commands": [
+        "pip3 install --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} kolibri",
+        "cp server.py /app/lib/python3.7/site-packages/kolibri/utils/server.py",
+        "cp cli.py /app/lib/python3.7/site-packages/kolibri/utils/cli.py"
```
Instead of override this files with the full copy maybe it's better to just have a diff file and patch. Something like this: https://github.com/learningequality/kolibri-installer-gnome/pull/15/files

dylanmccall committed:
Split KolibriDaemonProxy into a separate module

In addition, make better use of pygobject here.

dylanmccall committed:
Use KolibriDaemonProxy from the search provider

This replaces the search handler code that was running in the search
provider itself.

dylanmccall committed:
Use async method calls in KolibriDaemon

dylanmccall committed:
Improve lazy URL generating

Remove related code from KolibriDaemonProxy and simplify related code
paths.

dylanmccall committed:
Add config constants for well-known dbus object paths

dylanmccall committed:
Unown system bus name while shutting down

dylanmccall committed:
Use a shorter inactivity timeout if Kolibri is not running

dylanmccall committed:
Clean up KolibriSearchHandler's process pool on exit

dylanmccall committed:
Handle errors connecting to Kolibri Daemon

dylanmccall commented:
Here is some more detail about what is happening in this branch, which might be a useful basis for reviewing:

- The Kolibri service management code has been split into a separate application, kolibri-daemon. It works as a dbus service, with new service files included in `data/dbus`. The interface of this service is…
  - `Hold()`: Causes the service to keep running as long as the dbus client is connected.
  - `Release()`: Removes a previously created hold.
  - `Start()`: Starts Kolibri. Kolibri will keep running until the service stops.
  - `GetItemIdsForSearch()`: Search provider stuff.
  - `GetMetadataForItemIds()`: More search provider stuff.
  - `AppKey`: The unique app key for the front-end to use Kolibri in app mode.
  - `BaseURL`: The URL to access Kolibri, which may be different every time Kolibri starts.
  - `KolibriHome`: Kolibri's home directory.
  - `Status`: STOPPED, STARTING, STARTED, or ERROR.
  - `Version`: The version of kolibri-daemon's dbus interface. Currently 1, and we don't bother checking it, but it might be useful in the future.
- The code in kolibri-daemon is _mostly_ the same as what was there, but I included a patch for Kolibri which allows it to run on a random (system-assigned) port number and adds a callback so we know when it is ready to receive connections.
- I moved part of the search provider code to kolibri-daemon. The search provider itself doesn't need to load Kolibri or access its files: instead it asks kolibri-daemon to do the search stuff.
- Because the flatpak now has several entry points (the dbus daemon, a search provider, and the front-end), I tweaked the build system to better define the application ID for each of these (`base_application_id`, `frontend_application_id`, `daemon_application_id`, and `search_provider_application_id`), and added these variables to the config.py which is generated at build time instead of hard-coding stuff.
- I split some Python code that makes it easier to deal with GLib's dbus stuff out of the search provider component and into dbus-utils.py, since we are doing this in a few places now.
- I changed the desktop launcher code to talk to kolibri-daemon. This meant adding kolibri_daemon_proxy.py to represent kolibri-daemon (as a client), and changing some properties which were hard-coded or generated side-effects to instead be fetched over dbus.
- As a happy side-effect of this all, this ends up resolving issue #10 (logging has been kind of messed up due to different bits of Python stepping on each others' log configuration). The frontend (kolibri-gnome) is now just its own process so logging is easy there, and then for kolibri-daemon it's a well defined problem. I have it writing several log files (one for each of its subprocesses), which isn't _great_, but it does work.
- I added some files to build-aux/flatpak so we can open [GNOME Builder](https://wiki.gnome.org/Apps/Builder), clone this repository, and build the front-end (as a flatpak) very easily. Highly recommended :)

danigm reviewed (commented):
I've take a look to the code and looks good to me. I've left some comments.

I've been testing this and it's not working for me. The app is launched but it looks like the kolibri daemon service is not working correctly and the loading never ends.

dylanmccall commented on src/kolibri_gnome/kolibri_daemon/kolibri_search_handler.py:
```diff
@@ -51,8 +52,12 @@ def _item_id_to_node_id(item_id):
+            _kind_code, node_id = item_id.split("/", 1)
+        except ValueError:
+            raise ValueError("Invalid item_id")
+        else:
```
Haha, looking at that again, that is _very_ silly. I think I caught the ValueError thinking I'd like to raise a more specific exception, thought "hm, which built-in exception type should I use…" and ended up with … ValueError :b

dylanmccall committed:
Choose bus type based on environment in KolibriDaemonProxy

dylanmccall committed:
Clean up unnecessary try / except

dylanmccall commented:
Okay, I fixed that issue where the service wasn't connecting. I forgot to put back the code that gets it to connect over the session bus by default since I have been testing endlessm/eos-kolibri#1 in parallel with this :)

dylanmccall committed:
Use a command line argument for system or session mode

With this change, the bus type used by kolibri-daemon is determined with
a command line argument instead of an environment variable. This way, it
is possible to override KOLIBRI_USE_SYSTEM_INSTANCE when running the
Kolibri front-end, and have the corresponding dbus service start
correctly.

danigm reviewed (approved):
Good work, Now it's working correctly for me.

dylanmccall commented:
Okay, I'm merging this now. Next step will be a pull request for [flathub/org.learningequality.Kolibri](https://github.com/flathub/org.learningequality.Kolibri) :)
